File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.js
function solution(n, edge) {
    let graph = Array.from({length: n+1}, () => []);
    for(const [s, e] of edge){
        graph[s].push(e);
        graph[e].push(s);
    }
    
    const distance = Array(n+1).fill(0);
    const q = [1];
    const visited = [1];
    let head = 0;
    
    while(head < q.length){
        const cur = q[head++];
        
        for(const next of graph[cur]){
            if(!visited.includes(next)){
                visited.push(next);
                distance[next] = distance[cur] + 1
                q.push(next);
            }
        }
    }
    
    const maxDistance = Math.max(...distance);
    return distance.filter(d => d === maxDistance).length;
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    graph = [[] for _ in range(n+1)]
    for s, e in edge:
        graph[s].append(e)
        graph[e].append(s)
    
    distance = [0] * (n + 1)
    visited = [False] * (n + 1)
    q = deque([1])
    visited[1] = True
    
    while q:
        cur = q.popleft()
        for next in graph[cur]:
            if not visited[next]:
                visited[next] = True
                distance[next] = distance[cur] + 1
                q.append(next)
    maxDistance = max(distance)
    return distance.count(maxDistance)
                
        